src/Kernel/Process.class.st

```diff
@@ -218,13 +218,11 @@ Process >> doTerminationFromAnotherProcess [
 	It assumes that self is not the active process and the termination is requested from another process.
 	It assumes the terminator process won't resume before the process represented by self is terminated."	
 		
-	[ ] ensure: [ | oldList context terminator |
-		oldList := self suspend.
-		self isTerminated ifTrue: [^self].
+	[ ] ensure: [ | context terminator |
 		self isTerminating ifTrue: [ProcessAlreadyTerminating signal. ^self].
 		terminating := true.
```

Comment:
And I would move out of block this two lines together with nil check for #suspenedContext

Reply:
What do you mean exactly? Out of the ensure argument block? (That would break the case when the killer process is preempted before it enters the ensure argument block and then terminated - which would cause the argument block execute but theoretically the out-of-block checks could have been missed).

Reply:
I thought that moving the following checks would make the code inside the ensure a bit simpler (no returns for example):
```Smalltalk 
suspendedContext ifNil: [^self].
self isTerminating ifTrue: [ProcessAlreadyTerminating signal. ^self].

[] ensure: [...]
```

Reply:
Yes, it's true the chance the killer process gets preempted while sending #isTerminating is negligible. To be super careful I should have rather enclosed the new unwind suspendedContext's code into the ensure argument block:
`suspendedContext := [ [] ensure: [context unwindTo: nil. self endProcess] ] asContext.`
instead of plain:
`suspendedContext := [context unwindTo: nil. self endProcess] asContext.`
That could be the next iteration after I think it through :)

Reply:
Well, because in Pharo you have this `terminating` flag and `isTerminating` test the above extra-careful approach doesn't apply - you can't terminate again a process that is already flagged as 'terminating'. I'm not sure why you use the flag because it seems to me quite legitimate to terminate an already terminating process: e.g. if you terminate a process and the termination gets stuck (waiting on a resource etc.) you may want to reconsider and terminate it again. In Pharo it seems to me you just get a notification 'Process already terminating' and the remaining unwind blocks won't run. So WRT the above I'd leave the code as is for the moment :)